### PR TITLE
add automatic code format enforcer - black

### PR DESCRIPTION
**Pull request by gaborbernat**
``Black is the uncompromising Python code formatter. By using it, you agree to cede control over minutiae of hand-formatting. In return, Black gives you speed, determinism, and freedom from pycodestyle nagging about formatting. You will save time and mental energy for more important matters.`` - https://github.com/ambv/black

### CODE REVIEW

**Comment by obestwalter** [comment]
Hi @gaborbernat, are you using this project already somewhere else and can share some experiences with it? A first glimpse over the diff, looks like it is doing a good job. I also think I could get used to that style. Is it stable?

**Comment by obestwalter** [comment]
I can't say I like the style it enforces, especially how parentheses are always on their own lines, but I get the argument that it produces the smallest possible difffs and most of that is a matter of getting used to anyway.

http://black.readthedocs.io/en/stable/installation_and_usage.html#note-this-is-an-early-pre-release

It's marked alpha.

> Also, as a temporary safety measure, Black will check that the reformatted code still produces a valid AST that is equivalent to the original. This slows it down. If you’re feeling confident, use --fast.

... but that check puts at my worries to rest.

 I will start using it in my own code that I am working with for a while.

**Comment by gaborbernat** [comment]
It's not perfect, but I think it still provides much value.

**Comment by obestwalter** [comment]
I just read the https://github.com/ambv/black#the-black-code-style and I have to say that I am convinced.

I also agree with the thoughts about line length:

> You probably noticed the peculiar default line length. Black defaults to 88 characters per line, which happens to be 10% over 80. This number was found to produce significantly shorter files than sticking with 80 (the most popular), or even 79 (used by the standard library). In general, 90-ish seems like the wise choice.

So I would also propose to keep the black default for that.

**Comment by obestwalter** [comment]
Not a bug - missing (tricky) feature (string wrapping): https://github.com/ambv/black/issues/182

**Comment by gaborbernat** [comment]
Eh I like what we have with 100 now myself. As for the later, maybe down the line it will be fixed and then the formatter will fix that too automatically.

**Comment by obestwalter** [comment]
Irt does some really weird things with strings though ...

It turns

    fmt = (
        "%(asctime)s %(filename)s.%(funcName)s:%(lineno)d %(levelname)s: "
        "%(message)s"
    )

into 

    fmt = (
        "%(asctime)s %(filename)s.%(funcName)s:%(lineno)d %(levelname)s: " "%(message)s"
    )

It does not concatenate the strings again ...

**Comment by gaborbernat** [comment]
Hmmm may require some manual one time mangling then. Probably fill in an issue for it though.

**Comment by obestwalter** [comment]
See https://github.com/ambv/black/issues/205

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -1,13 +1,19 @@
...
-    sha: v1.1.1
+    rev: v1.2.3
     hooks:
     -   id: trailing-whitespace
     -   id: end-of-file-fixer
     -   id: check-yaml
     -   id: debug-statements
     -   id: flake8
+-   repo: https://github.com/ambv/black
+    rev: stable
```
[the remark this answers is not in the document]
Comment (gaborbernat) [reply .pre-commit-config.yaml]: fixed, and yeah I agree about the quotes; but at the end of day I prefer consistency over which of the two :-)

**Comment by gaborbernat** [comment]
@obestwalter take another look cleaned up manually few of the oddities, if there are any more you spot can do similarly

**Comment by asottile** [comment]
Another option would be to use autopep8 + double-quote-string-fixer (converts strings in the opposite directing of black) + a few other one off formatters to get basically the same deal :man_shrugging:

**Comment by gaborbernat** [comment]
I'm personally not that pedantic about the single quote to consider worth the effort there.

**Comment by gaborbernat** [comment]
@obestwalter it now passed tests, let me know if you spot anything else :smile:

**Comment by obestwalter** [comment]
> I wish it was single quotes instead :/

Hey @asottile, if you are interested how this came to pass: I just stumbled over the issue dealing with this: https://github.com/ambv/black/issues/51 - there are a few very strong arguments why double quotes are the better choice. The TL;DR of the tipping point is: apostrophes (https://github.com/ambv/black/pull/75#issuecomment-376193551).

Regarding other tools: the more I look into black the more I get the feeling that this could become a standard that will save the community a lot of bike shedding.

I really like the idea to approach the problem from the angle that code should be formatted in a way that maintainability and preotection from stupid bugs is prioritized. For that goal it is doing stuff that "traditional" formatters won't touch and IMO that is a GoodThing. Prominently protecting us from the infamous: "Whoops there should actually have been a comma between those two strings in a list, where I moved an element from the end to the middle". 

@ambv also has a very clear vision, but is open to change his opinion if confronted with strong arguments (e.g. he also preferred single quotes first), so this looks like a very good project to go with and thanks to the AST checking safeguard, we do not have to be too worried that this first big reformatting PR will introduce some subtle bugs.

The only problems I see atm are around the topic of strings, but they seem to be not too painful and will likely get better. I noticed that our way of creating these mini configs with `initproj` gets totally shot, which IMO is another black bug, as it actually changes the formatting of the string that will be outputted. In our case it is only an annoyance as the change in formatting does not affect the parsed config. It just looks really ugly now, but they can be formatted in normal strings anyway (I don't need a triple quoted string to create a test config with one or two line breaks), so I would suggest to reformat simple configs like:

        config = newconfig(
            [],
            """
            [testenv:py1]
            deps=pytest
        """,
        )

to:

config = newconfig([], "[testenv:py1]\ndeps=pytest")

Which is much more readable than even the default formatting anyway.

I am going through the PR today and if there are no objections, I think we should whip this into shape together quickly as long as @gaborbernat has still his PyCon wings on :)

**Comment by asottile** [comment]
> the more I look into black the more I get the feeling that this could
become a standard that will save the community a lot of bike shedding.

yapf started with exactly the same mentality, and is much more mature.

> For that goal it is doing stuff that "traditional" formatters won't touch

true yapf / autopep8 don't even try here but that isn't to say that all
formatters don't:

- https://github.com/asottile/add-trailing-comma
- https://github.com/asottile/pyupgrade

> The TL;DR of the tipping point is: apostrophes

For me it's less of a personal preference thing and more of an annoyance.
A good chunk of the python I write is for web development where double
quotes in strings are extremely common (html!).  The linked comment above
happens way more frequently in the _other direction_ for me.  It's also
contrary to nearly every style guide I've worked with (again, because web
development, but also because pressing the shift key is a slow and `"ERorr
prone"` operation).

> config = newconfig([], "[testenv:py1]\ndeps=pytest")

> Which is much more readable than even the default formatting anyway.

Having used this corner-cut myself before, I find it all too easy to typo
`\b` instead of `\n` whereas with multiline strings this isn't a problem.
Another note in your example is the two aren't equivalent -- there's a
whitespace difference (`textrwap.dedent` comes to mind).

Anyway, I'm not out here to change minds -- code formatters are some of my
favorite things and I think using more of them is a great step in the right
direction.  Just wish @ambv's black didn't strongly prefer (in my mind) the
less convenient quotation style.



On Sun, May 13, 2018 at 3:04 AM, Oliver Bestwalter <notifications@github.com

**Comment by gaborbernat** [comment]
Given ``tox`` is definitely not a web tool, I can get along with having ``"``. As for the initproj we already dedent by default the code, so I think that's just fine.

**Comment by RonnyPfannschmidt** [comment]
i just finished applying black to 2 of my projects along with pre-commit hooks

while my quoting preference is also different, i consider it a straw-man argument, you fix it once, and then keep it consistent -> and consistency is more important there imho

**Comment by obestwalter** [comment]
I understand your annoyance @asottile - it is very dependent on in which kind of universe your code lives - maybe that should really be something configurable in black and maybe it will be at some point to get those kind of projects on board (line lenght is configurable already after all). 

The way how black deals with quotes is definitely quite sane IMO - independent of preferences. It always tries to reformat the string in a way that you need the least amount of escapes. e.g. It and leaves strings like `bla = 'hello "world"'` alone and reformats strings like `blu = "hello \"world\""` to blu = 'hello "world"'

As you seem o.k. with using black anyway as a step in the right direction and we have an implicit +1 from @ronnypfannschmidt also, I would say let's just run with this for now. 

I'll finish reviewing this tonight then and merge it, so that new PRs coming in tomorrow at the sprint can already base on this, o.k.?

File: tox/session.py [hunk 2]
```diff
@@ -59,24 +59,29 @@ def show_help(config):
     tw.write(config._parser._format_help())
     tw.line()
     tw.line("Environment variables", bold=True)
-    tw.line("TOXENV: comma separated list of environments "
-            "(overridable by '-e')")
-    tw.line("TOX_TESTENV_PASSENV: space-separated list of extra "
-            "environment variables to be passed into test command "
-            "environments")
+    tw.line("TOXENV: comma separated list of environments " "(overridable by '-e')")
```
Comment (obestwalter) [inline tox/session.py]: needs manual concat.

File: tox/venv.py [hunk 3]
```diff
@@ -375,13 +377,21 @@ def test(self, redirect=False):
...
+                        ignore_ret=ignore_ret,
+                        testcommand=True,
+                    )
                 except tox.exception.InvocationError as err:
                     if self.envconfig.ignore_outcome:
                         self.session.report.warning(
-                            "command failed but result from testenv is ignored\n"
-                            "  cmd: %s" % (str(err),))
+                            "command failed but result from testenv is ignored\ncmd: {}".format(
+                                str(err)
```
Comment (obestwalter) [inline tox/venv.py]: loose the str() here.

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -1,13 +1,19 @@
...
+    rev: 18.4a4
+    hooks:
+    - id: black
+      args: [--line-length=99, --safe]
+      python_version: python3.6
 -   repo: https://github.com/pre-commit/pre-commit-hooks
-    sha: v1.1.1
+    rev: v1.2.3
     hooks:
     -   id: trailing-whitespace
```
Comment (obestwalter) [inline .pre-commit-config.yaml]: Not necessary with black anymore, I guess.

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -1,13 +1,19 @@
...
     -   id: end-of-file-fixer
```
Comment (obestwalter) [inline .pre-commit-config.yaml]: see above

File: README.rst [hunk 6]
```diff
@@ -16,6 +16,9 @@
 .. image:: https://readthedocs.org/projects/tox/badge/?version=latest
   :target: http://tox.readthedocs.io/en/latest/?badge=latest
   :alt: Documentation status
+.. image:: https://img.shields.io/badge/code%20style-black-000000.svg
+  :target: https://github.com/ambv/black
+  :alt: Code style: black
```
Comment (obestwalter) [inline README.rst]: Yeah, I guess that is the case now :)

File: tests/test_config.py [hunk 7]
```diff
@@ -8,27 +8,33 @@
...
+
     def test_config_parsing_minimal(self, tmpdir, newconfig):
-        config = newconfig([], """
+        config = newconfig(
+            [],
+            """
             [testenv:py1]
-        """)
+        """,
+        )
```
[the remark this answers is not in the document]
Comment (gaborbernat) [reply tests/test_config.py]: Yeah, the comma in this case is optional, but never changes the meaning.

[the remark this answers is not in the document]
Comment (obestwalter) [reply tests/test_config.py]: I was rather thinking that the line break introduced after the `"""` would introduce changes in formatting (in some cases - not this one, so the comment was placed not in a good example), but they don't.

File: tests/test_pytest_plugins.py [hunk 8]
```diff
@@ -10,89 +10,97 @@
...
+        "input, expected",
+        (
+            ("", []),
+            ("/", ["/"]),
+            ("//", ["//"]),
+            ("/a", ["/", "a"]),
+            ("/a/", ["/", "a"]),
+            ("/a/b", ["/", "a", "b"]),
+            ("a", ["a"]),
+            ("a/b", ["a", "b"]),
```
[the remark this answers is not in the document]
Comment (gaborbernat) [reply tests/test_pytest_plugins.py]: indeed :laughing:

File: tests/test_result.py [hunk 9]
```diff
@@ -65,31 +69,34 @@ def test_get_commandlog(pkg):
     assert "setup" not in envlog.dict
     setuplog = envlog.get_commandlog("setup")
     setuplog.add_command(["virtualenv", "..."], "venv created", 0)
-    assert setuplog.list == [{"command": ["virtualenv", "..."],
-                              "output": "venv created",
-                              "retcode": "0"}]
+    assert (
+        setuplog.list
+        == [{"command": ["virtualenv", "..."], "output": "venv created", "retcode": "0"}]
```
[the remark this answers is not in the document]
Comment (gaborbernat) [reply tests/test_result.py]: removed all these by extracting the constant to a local variable

**Comment by obestwalter** [comment]
@asottile said:
>> config = newconfig([], "[testenv:py1]\ndeps=pytest")
>> Which is much more readable than even the default formatting anyway.
> Having used this corner-cut myself before, I find it all too easy to typo
> `\b` instead of `\n` whereas with multiline strings this isn't a problem.
> Another note in your example is the two aren't equivalent -- there's a
> whitespace difference (`textrwap.dedent` comes to mind).

In general I agree, but I think in the context of a tox config test this is not a problem. If you accidentally type `\b` your test will fail and you'll fix the typo. The whitespace difference is also irrelevant in the context of those particular tests.

**Comment by gaborbernat** [comment]
@obestwalter should be ready for merge now, :smile: as for doing the ``%`` to ``str.format`` I did it mostly by search, and pycharm suggestions

**Comment by obestwalter** [comment]
> as for doing the % to str.format I did it mostly by search, and pycharm suggestions

Wow. You've been busy then :) thanks for all the work you put in to whip that into shape.

File: tests/test_config.py [hunk 10]
```diff
@@ -8,9 +8,16 @@
...
+from tox.config import (
+    CommandParser,
+    DepOption,
+    SectionReader,
+    get_homedir,
+    get_version_info,
+    getcontextname,
+    is_section_substitution,
+    parseconfig,
+)
```
Comment (obestwalter) [inline tests/test_config.py]: argh, now we're back to that thing, but tbh atm I don't care anymore. From now on it's just a question of reconfiguring whatever formatter we use.

File: tests/test_config.py [hunk 11]
```diff
@@ -2659,28 +2647,17 @@ def test_command_with_split_line_in_subst_arguments(self):
...
-                " ",
-                "!deferred",
-                " ",
-                "--with-doctest",
-                " ",
-                "[]",
-            ]
-        )
+        expected = [
+            "nosetests", " ", "-v", " ", "-a", " ", "!deferred", " ", "--with-doctest", " ", "[]"
```
Comment (obestwalter) [inline tests/test_config.py]: phew - thanks.

File: tests/test_config.py (hunk 10 above)
Comment (gaborbernat) [reply tests/test_config.py]: yeah, maybe things will change in future with https://github.com/ambv/black/issues/204

Comment (gaborbernat) [reply tests/test_config.py]: stays as it is then :+1:

**Pull request merged**